kotlin-codepoints-deluxe/src/commonMain/kotlin/CodePoint.kt

```diff
@@ -0,0 +1,129 @@
+package de.cketti.codepoints.deluxe
+
+import de.cketti.codepoints.CodePoints
+import kotlin.jvm.JvmInline
+
+/**
+ * Represents a Unicode code point.
+ * 
+ * You can create/retrieve instances of this class by using the following functions:
+ * - [Int.toCodePoint]
+ * - [Char.toCodePoint]
+ * - [String.codePointSequence]
+ * - [String.codePointIterator]
+ */
+@JvmInline
+value class CodePoint internal constructor(val value: Int) {
+    init {
+        require(CodePoints.isValidCodePoint(value)) { "Not a valid code point" }
+    }
+
+    /**
+     * `true` if this Unicode code point is in the Basic Multilingual Plane (BMP).
+     *
+     * Such code points can be represented using a single `Char`.
+     */
+    val isBasic: Boolean
+        get() = CodePoints.isBmpCodePoint(value)
+
+    /**
+     * `true` if this Unicode code point is in the supplementary character range.
+     *
+     * In a `String` such a code point is represented using a surrogate pair, i.e. two `Char` values.
+     */
+    val isSupplementary: Boolean
+        get() = CodePoints.isSupplementaryCodePoint(value)
+
+    /**
+     * The number of `Char` values needed to represent this Unicode code point.
+     *
+     * If the specified code point is in the [BMP][CodePoint.isBasic], this property is `1`, otherwise it is `2`.
+     */
+    val charCount: Int
+        get() = CodePoints.charCount(value)
+
+    /**
+     * `true` if this code point is a surrogate code unit.
+     */
+    val isSurrogate: Boolean
+        get() = !isSupplementary && value.toChar().isSurrogate()
+
+    /**
+     * `true` if this code point is a high surrogate code unit.
+     */
+    val isHighSurrogate: Boolean
+        get() = !isSupplementary && value.toChar().isHighSurrogate()
+
+    /**
+     * `true` if this code point is a low surrogate code unit.
+     */
+    val isLowSurrogate: Boolean
+        get() = !isSupplementary && value.toChar().isLowSurrogate()
```

Comment:
What would these be used for? They seem to go against my intuition of what a code point is since I thought they were exclusively properties of characters.

Reply:
I think of the value of a code point as the index into the Unicode data table. Surrogate characters are part of that table, too. 

Most of the time we don't care about surrogate pairs and want to get the code point encoded by the two surrogate code points. But occasionally you might be interested in dealing with surrogate characters on their own. Maybe we're missing a `combineWith()` method that returns the value of `CodePoints.toCodePoint(high, low)` when called like `high.combineWith(low)`.

Reply:
Good to know!